Prevent the CheckPointer from blocking CausalCluster Catchup.

chrisvest opened the pull request:
By making the checkpointer not hold on to the PageCache monitor lock
for the whole check point process.

It turned out that this behaviour could inconvenience the Causal Cluster
catchup process.
The CC catchup would ask the page cache warmer to list its files.
To do that, the page cache warmer would call `listExistingMappings` on
the page cache, and that method would also require the PageCache lock.

The fix is that the checkpointer now also uses `listExistingMappings`
to grab a snapshot of the mapped files, and then flush them without
holding on to the lock on the entire page cache.
The `listExistingMappings` method does still take the page cache lock,
but it returns much, much faster than a complete page cache flush would.

chrisvest committed:
Prevent the CheckPointer from blocking CausalCluster Catchup.

By making the checkpointer not hold on to the PageCache monitor lock
for the whole check point process.

It turned out that this behaviour could inconvenience the Causal Cluster
catchup process.
The CC catchup would ask the page cache warmer to list its files.
To do that, the page cache warmer would call `listExistingMappings` on
the page cache, and that method would also require the PageCache lock.

The fix is that the checkpointer now also uses `listExistingMappings`
to grab a snapshot of the mapped files, and then flush them without
holding on to the lock on the entire page cache.
The `listExistingMappings` method does still take the page cache lock,
but it returns much, much faster than a complete page cache flush would.

chrisvest committed:
Add a test that verifies that `PageCache.flushAndForce` does not hold the `PageCache` lock the hole time.

klaren commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCache.java:
```diff
@@ -573,30 +573,30 @@ public void flushAndForce() throws IOException
         }
         assertNotClosed();
-        flushAllPages( limiter );
+        List<PagedFile> files = listExistingMappings();
```
How is the `calling code must be prepared for that the returned PagedFiles can be asynchronously closed elsewhere.` handled? `org.neo4j.io.pagecache.impl.muninn.MuninnPagedFile#flushAndForceInternal` does not validate that the file is still mapped as far as I can tell.

klaren replied:
The failing tests appear to be related to this:
```
Caused by: org.neo4j.kernel.impl.store.UnderlyingStorageException: Failed to flush
Caused by: java.nio.channels.ClosedChannelException
```

chrisvest replied:
Yes, see the commit I just pushed. It's technically _"fine"_ because the check pointer will retry, but it is obviously inconveniencing the tests. It also pointed to some optimisation opportunities that @burqen will look at in due time.

chrisvest committed:
Make PageCache.flushAndForce cope with concurrent file unmappings.

This is important because `flushAndForce` is called during steady state as part of check point.
Meanwhile, mapping and unmapping files are also happening during steady state operations when indexes are created or dropped.
And now that `flushAndForce` is no longer holding the `PageCache` monitor lock throughout the whole thing, those two operations can happen concurrently.

klaren reviewed (approved):
Looks good now!
